fili-core/src/main/java/com/yahoo/bard/webservice/web/endpoints/JobsServlet.java

```diff
@@ -359,20 +355,50 @@ protected void handlePreResponse(
+     * @param asyncAfter  The minimum duration the request is allowed to last before becoming asynchronous
+     *
+     * @return An Observable wrapping a PreResponse or an empty Observable in case a timeout occurs.
+     */
+    protected Observable<PreResponse> getResults(@NotNull String ticket, long asyncAfter) {
+        /*
+         * BroadCastChannel is a hot observable i.e. it emits notification irrespective of whether it has any
+         * subscribers. We use the replay operator so that the preResponseObservable upon connection, will begin
+         * collecting values.
+         * Once a new observer subscribes to the observable, it will have all the collected values replayed to it.
+         */
+        if (asyncAfter == JobsApiRequest.ASYNCHRONOUS_ASYNC_AFTER_VALUE) {
+            return Observable.empty();
+        } else {
+            ConnectableObservable<String> broadcastChannelNotifications = broadcastChannel.getNotifications()
+                    .filter(ticket::equals)
+                    .take(1)
+                    .replay(1);
+            broadcastChannelNotifications.connect();
+            return preResponseStore.get(ticket).switchIfEmpty(
+                    applyTimeoutIfNeeded(broadcastChannelNotifications, asyncAfter).flatMap(preResponseStore::get)
+            );
+        }
+    }
+
+    /**
+     * Given an observable, returns a new observable with an asyncAfter timeout applied only if {@code asyncAfter} is
+     * not {@code never}.
+     * <p>
+     * If the timeout expires, the current observable is replaced with an empty observable.
+     *
+     * @param observable  The observable that should have a timeout attached to it, if the request's asyncAfter is
+     * a number
+     * @param asyncAfter  The minimum duration the request is allowed to last before becoming asynchronous
+     * @param <T>  The type of the observable's payload
      *
-     * @return An Observable over PreResponse or an empty Observable
+     * @return An Observable that may or may not have a timeout attached to it, depending on whether the request is
+     * forced to be synchronous or not
      */
-    protected Observable<PreResponse> getPreResponseObservable(String ticket, JobsApiRequest apiRequest) {
-        //Subscribe to the BroadcastChannel so we do not miss notifications about the ticket being stored in the
-        //PreResponseStore.
-        Observable<PreResponse> broadCastChannelPreResponseObservable =
-                apiRequest.handleBroadcastChannelNotification(ticket);
-
-         //Check the PreResponseStore to see if the PreResponse associated with the given ticket has been stored in
-        //the PreResponseStore. If not, wait for a PreResponse for the amount of time specified in async.
-        return preResponseStore.get(ticket).switchIfEmpty(broadCastChannelPreResponseObservable);
+    private <T> Observable<T> applyTimeoutIfNeeded(Observable<T> observable, long asyncAfter) {
```

Comment:
If you look closely, you'll see that the result of `applyTimeOutIfNeeded` is actually being applied _inside_ of a `preResponseStore.get().switchIfEmpty` call. In other words, we aren't checking the PreResponseStore if the request is always asynchronous (there's no point), but we are in the other two cases.

So we can't really merge the three cases into a single helper method without more thought than is worth it. If we did, we'd probably just end up with implementing all of `getResults` in `applyTimeoutIfNeeded`.